### PR TITLE
[ML] Attend to outstanding Boost.JSON tasks.

### PR DESCRIPTION
**Pull request by edsavage**
Attend to those non-essential, nice-to-have items related to the switch to Boost.JSON

relates #2622

### CODE REVIEW

File: lib/api/CNdJsonInputParser.cc [hunk 1]
```diff
@@ -10,6 +10,7 @@
  */
 #include <api/CNdJsonInputParser.h>
 
+#include "core/CBoostJsonParser.h"
```
Comment (droberts195) [inline lib/api/CNdJsonInputParser.cc:13]:
```suggestion
#include <core/CBoostJsonParser.h>
```

File: lib/api/CRetrainableModelJsonReader.cc [hunk 2]
```diff
@@ -26,6 +26,7 @@
 // using basic_parser to implement a parser.
 #include <boost/json/basic_parser_impl.hpp>
 
+#include "core/CBoostJsonParser.h"
```
Comment (droberts195) [inline lib/api/CRetrainableModelJsonReader.cc:29]: Use angle brackets and move with other `core` imports.

File: lib/api/CSerializableToJson.cc [hunk 3]
```diff
@@ -23,6 +23,7 @@
 #include <boost/iostreams/filtering_stream.hpp>
 #include <boost/iostreams/stream.hpp>
 
+#include "core/CBoostJsonParser.h"
```
Comment (droberts195) [inline lib/api/CSerializableToJson.cc:26]: Use angle brackets and move with other `core` imports.

File: lib/maths/analytics/CDataFrameCategoryEncoder.cc [hunk 4]
```diff
@@ -239,6 +231,74 @@ class CMinRedundancyMaxRelevancyGreedySearch {
...
+        err << encodingObj.as_object().at(tag);
+        throw std::runtime_error("JSON value is not an object: " + err.str());
+    }
+    const json::object& obj = encodingObj.as_object().at(tag).as_object();
+    LOG_TRACE(<< tag << " obj: " << obj);
+
+    getEncodingAttribute(obj, ENCODING_INPUT_COLUMN_INDEX_TAG, colIdx);
+    getEncodingAttribute(obj, ENCODING_MIC_TAG, mic);
+
+    return std::cref(obj);
```
Comment (droberts195) [inline lib/maths/analytics/CDataFrameCategoryEncoder.cc:268]: Why is `std::cref` needed here?

File: lib/maths/analytics/CDataFrameCategoryEncoder.cc [hunk 5]
```diff
@@ -239,6 +231,74 @@ class CMinRedundancyMaxRelevancyGreedySearch {
...
+void extractMappedEncodingAndFallback(const json::object& obj, TDoubleVec& map, double& fallback) {
+    std::ostringstream err;
+    const json::value& encodingMap = obj.at(MAPPED_ENCODING_MAP_TAG);
+    if (encodingMap.is_string() == false) {
+        err << encodingMap;
+        throw std::runtime_error("JSON value is not a string: " + err.str());
+    }
+    core::CStringUtils::TStrVec tokens;
+    std::string rmdr;
+    core::CStringUtils::tokenise(":", encodingMap.as_string().c_str(), tokens, rmdr);
```
Comment (droberts195) [inline lib/maths/analytics/CDataFrameCategoryEncoder.cc:280]: Longer term (not necessary for this PR) we should make `CStringUtils::tokenise` take a `std::string_view` as input. That would avoid the inefficiency here of creating a second `std::string` in addition to Boost.JSON's "stringlike" object.

Comment (edsavage) [reply lib/maths/analytics/CDataFrameCategoryEncoder.cc:280]: Noted

**Review by droberts195 — APPROVED** [review]
LGTM if CI passes

**Pull request merged**